src/main/flight/imu.c

```diff
@@ -174,126 +204,222 @@ void imuCalculateAcceleration(uint32_t deltaT)
-    // or handle the case in which they are and (atan2f(0, 0) is undefined.
-    float hd = (atan2f(Yh, Xh) * 1800.0f / M_PIf + magneticDeclination) / 10.0f;
-    head = lrintf(hd);
-
-    // Arctan returns a value in the range -180 to 180 degrees. We 'normalize' negative angles to be positive.
-    if (head < 0)
-        head += 360;
-
-    return head;
+    return 1.0f / sqrtf(x);
 }
 
-static void imuCalculateEstimatedAttitude(void)
+static bool imuUseFastGains(void)
 {
-    int32_t axis;
-    int32_t accMag = 0;
-    static t_fp_vector EstM;
-    static t_fp_vector EstN = { .A = { 1.0f, 0.0f, 0.0f } };
-    static float accLPF[3];
-    static uint32_t previousT;
-    uint32_t currentT = micros();
-    uint32_t deltaT;
-    float scale;
-    fp_angles_t deltaGyroAngle;
-    deltaT = currentT - previousT;
-    scale = deltaT * gyroScaleRad;
-    previousT = currentT;
-
-    // Initialization
-    for (axis = 0; axis < 3; axis++) {
-        deltaGyroAngle.raw[axis] = gyroADC[axis] * scale;
-        if (imuRuntimeConfig->acc_lpf_factor > 0) {
-            accLPF[axis] = accLPF[axis] * (1.0f - (1.0f / imuRuntimeConfig->acc_lpf_factor)) + accADC[axis] * (1.0f / imuRuntimeConfig->acc_lpf_factor);
-            accSmooth[axis] = accLPF[axis];
-        } else {
-            accSmooth[axis] = accADC[axis];
-        }
-        accMag += (int32_t)accSmooth[axis] * accSmooth[axis];
+    return !ARMING_FLAG(ARMED) && millis() < 20000;
+}
+
+static float imuGetPGainScaleFactor(void)
+{
+    if (imuUseFastGains()) {
+        return 10.0f;
+    }
+    else {
+        return 1.0f;
     }
-    accMag = accMag * 100 / ((int32_t)acc_1G * acc_1G);
+}
 
-    rotateV(&EstG.V, &deltaGyroAngle);
+static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
+                                bool useAcc, float ax, float ay, float az,
+                                bool useMag, float mx, float my, float mz,
+                                bool useYaw, float yawError)
+{
+    static float integralFBx = 0.0f,  integralFBy = 0.0f, integralFBz = 0.0f;    // integral error terms scaled by Ki
+    float recipNorm;
+    float hx, hy, bx;
+    float ex = 0, ey = 0, ez = 0;
+    float qa, qb, qc;
+
+    // Calculate general spin rate (rad/s)
+    float spin_rate = sqrtf(sq(gx) + sq(gy) + sq(gz));
+
+    // Use raw heading error (from GPS or whatever else)
+    if (useYaw) {
+        while (yawError >  M_PIf) yawError -= (2.0f * M_PIf);
+        while (yawError < -M_PIf) yawError += (2.0f * M_PIf);
+
+        ez += sin_approx(yawError / 2.0f);
+    }
 
-    // Apply complimentary filter (Gyro drift correction)
-    // If accel magnitude >1.15G or <0.85G and ACC vector outside of the limit range => we neutralize the effect of accelerometers in the angle estimation.
-    // To do that, we just skip filter, as EstV already rotated by Gyro
+    // Use measured magnetic field vector
+    recipNorm = mx * mx + my * my + mz * mz;
```

Comment:
I guess sq() could be used here (and everywhere square is computed) like on line 239